File: packages/udentify-core-flutter/android/src/main/kotlin/com/udentifycoreflutter/SSLPinningManager.kt
package com.udentifycoreflutter

import android.content.Context
import android.util.Base64
import android.util.Log
import io.udentify.android.commons.model.UdentifySettingsProvider
import java.security.cert.CertificateFactory
import java.security.cert.X509Certificate

/**
 * SSLPinningManager
 * Manages SSL certificate pinning using UdentifySettingsProvider
 */
class SSLPinningManager(private val context: Context) {
    
    companion object {
        private const val TAG = "SSLPinningManager"
    }
    
    /**
     * Load a certificate from the Android assets folder and set it for SSL pinning
     */
    fun loadCertificateFromAssets(
        certificateName: String,
        extension: String,
        onSuccess: () -> Unit,
        onError: (String) -> Unit
    ) {
        try {
            Log.d(TAG, "loadCertificateFromAssets called: $certificateName.$extension")
            
            // Use UdentifySettingsProvider to load the certificate
            val certificate = UdentifySettingsProvider.loadDERCertificateData(
                context,
                certificateName,
                extension
            )
            
            if (certificate == null) {
                val errorMessage = "Failed to load certificate '$certificateName.$extension' from assets. Ensure the certificate file exists in the assets folder and is in DER format."
                Log.e(TAG, "Error: $errorMessage")
                onError(errorMessage)
                return
            }
            
            Log.d(TAG, "Certificate loaded, subject: ${certificate.subjectDN}")
            
            // Set the certificate using UdentifySettingsProvider
            UdentifySettingsProvider.setSSLCertificate(certificate)
            
            Log.d(TAG, "Certificate set successfully")
            onSuccess()
            
        } catch (e: Exception) {
            Log.e(TAG, "Exception: ${e.message}", e)
            onError(e.message ?: "Unknown error occurred")
        }
    }
    
    /**
     * Set SSL certificate using base64 encoded data
     */
    fun setSSLCertificateBase64(
        certificateBase64: String,
        onSuccess: () -> Unit,
        onError: (String) -> Unit
    ) {
        try {
            Log.d(TAG, "setSSLCertificateBase64 called")
            
            // Decode base64 string
            val certificateBytes = Base64.decode(certificateBase64, Base64.DEFAULT)
            
            Log.d(TAG, "Certificate data decoded, size: ${certificateBytes.size} bytes")
            
            // Convert bytes to X509Certificate
            val certificateFactory = CertificateFactory.getInstance("X.509")
            val certificate = certificateFactory.generateCertificate(
                certificateBytes.inputStream()
            ) as X509Certificate
            
            Log.d(TAG, "Certificate parsed, subject: ${certificate.subjectDN}")
            
            // Set the certificate using UdentifySettingsProvider
            UdentifySettingsProvider.setSSLCertificate(certificate)
            
            Log.d(TAG, "Certificate set successfully")
            onSuccess()
            
        } catch (e: Exception) {
            Log.e(TAG, "Exception: ${e.message}", e)
            onError(e.message ?: "Unknown error occurred")
        }
    }
    
    /**
     * Remove the currently set SSL certificate
     */
    fun removeSSLCertificate(
        onSuccess: () -> Unit,
        onError: (String) -> Unit
    ) {
        try {
            Log.d(TAG, "removeSSLCertificate called")
            
            // Remove certificate using UdentifySettingsProvider
            UdentifySettingsProvider.removeSSLCertificate()
            
            Log.d(TAG, "Certificate removed successfully")
            onSuccess()
            
        } catch (e: Exception) {
            Log.e(TAG, "Exception: ${e.message}", e)
            onError(e.message ?: "Unknown error occurred")
        }
    }
    
    /**
     * Get the currently set SSL certificate as base64 string
     */
    fun getSSLCertificateBase64(
        onSuccess: (String?) -> Unit,
        onError: (String) -> Unit
    ) {
        try {
            Log.d(TAG, "getSSLCertificateBase64 called")
            
            // Get certificate using UdentifySettingsProvider
            val certificate = UdentifySettingsProvider.getSSLCertificate()
            
            if (certificate == null) {
                Log.d(TAG, "No certificate is currently set")
                onSuccess(null)
                return
            }
            
            // Convert certificate to base64
            val certificateBytes = certificate.encoded
            val base64String = Base64.encodeToString(certificateBytes, Base64.NO_WRAP)
            
            Log.d(TAG, "Certificate retrieved, size: ${certificateBytes.size} bytes")
            onSuccess(base64String)
            
        } catch (e: Exception) {
            Log.e(TAG, "Exception: ${e.message}", e)
            onError(e.message ?: "Unknown error occurred")
        }
    }
    
    /**
     * Check if SSL pinning is enabled
     */
    fun isSSLPinningEnabled(
        onSuccess: (Boolean) -> Unit,
        onError: (String) -> Unit
    ) {
        try {
            Log.d(TAG, "isSSLPinningEnabled called")
            
            // Check SSL pinning status using UdentifySettingsProvider
            val isEnabled = UdentifySettingsProvider.isSSLPinningEnabled()
            
            Log.d(TAG, "SSL pinning enabled: $isEnabled")
            onSuccess(isEnabled)
            
        } catch (e: Exception) {
            Log.e(TAG, "Exception: ${e.message}", e)
            onError(e.message ?: "Unknown error occurred")
        }
    }
}

File: packages/udentify-core-flutter/ios/Classes/SSLPinningManager.swift
import Foundation
import UdentifyCommons

/**
 * SSLPinningManager
 * Manages SSL certificate pinning using UdentifySettingsProvider
 */
@objc(SSLPinningManager)
public class SSLPinningManager: NSObject {
    
    /**
     * Load a certificate from the app bundle and set it for SSL pinning
     */
    @objc public func loadCertificateFromAssets(
        _ certificateName: String,
        extension fileExtension: String,
        completion: @escaping (Bool, Error?) -> Void
    ) {
        NSLog("SSLPinningManager - Loading certificate: \(certificateName).\(fileExtension)")
        
        // Load certificate data from bundle using UdentifySettingsProvider
        guard let certificateData = UdentifySettingsProvider.loadDERCertificateData(
            from: Bundle.main,
            certificateName: certificateName,
            certificateExtension: fileExtension
        ) else {
            let error = NSError(
                domain: "com.udentify.sslpinning",
                code: 1,
                userInfo: [NSLocalizedDescriptionKey: "Failed to load certificate '\(certificateName).\(fileExtension)' from bundle. Ensure the certificate file exists in the app bundle and is in DER format."]
            )
            NSLog("SSLPinningManager - Error: \(error.localizedDescription)")
            completion(false, error)
            return
        }
        
        NSLog("SSLPinningManager - Certificate data loaded, size: \(certificateData.count) bytes")
        
        // Set the certificate using UdentifySettingsProvider
        UdentifySettingsProvider.setSSLCertificate(with: certificateData)
        
        NSLog("SSLPinningManager - Certificate set successfully")
        completion(true, nil)
    }
    
    /**
     * Set SSL certificate using base64 encoded data
     */
    @objc public func setSSLCertificateBase64(
        _ certificateBase64: String,
        completion: @escaping (Bool, Error?) -> Void
    ) {
        NSLog("SSLPinningManager - Setting certificate from base64 data")
        
        // Decode base64 string to Data
        guard let certificateData = Data(base64Encoded: certificateBase64) else {
            let error = NSError(
                domain: "com.udentify.sslpinning",
                code: 2,
                userInfo: [NSLocalizedDescriptionKey: "Invalid base64 encoded certificate data"]
            )
            NSLog("SSLPinningManager - Error: \(error.localizedDescription)")
            completion(false, error)
            return
        }
        
        NSLog("SSLPinningManager - Certificate data decoded, size: \(certificateData.count) bytes")
        
        // Set the certificate using UdentifySettingsProvider
        UdentifySettingsProvider.setSSLCertificate(with: certificateData)
        
        NSLog("SSLPinningManager - Certificate set successfully")
        completion(true, nil)
    }
    
    /**
     * Remove the currently set SSL certificate
     */
    @objc public func removeSSLCertificate(
        _ completion: @escaping (Bool, Error?) -> Void
    ) {
        NSLog("SSLPinningManager - Removing SSL certificate")
        
        // Remove certificate using UdentifySettingsProvider
        UdentifySettingsProvider.removeSSLCertificate()
        
        NSLog("SSLPinningManager - Certificate removed successfully")
        completion(true, nil)
    }
    
    /**
     * Get the currently set SSL certificate as base64 string
     */
    @objc public func getSSLCertificateBase64(
        _ completion: @escaping (String?, Error?) -> Void
    ) {
        NSLog("SSLPinningManager - Getting SSL certificate")
        
        // Get certificate data using UdentifySettingsProvider
        guard let certificateData = UdentifySettingsProvider.getSSLCertificate() else {
            NSLog("SSLPinningManager - No certificate is currently set")
            completion(nil, nil)
            return
        }
        
        // Convert to base64 string
        let base64String = certificateData.base64EncodedString()
        NSLog("SSLPinningManager - Certificate retrieved, size: \(certificateData.count) bytes")
        
        completion(base64String, nil)
    }
    
    /**
     * Check if SSL pinning is enabled
     */
    @objc public func isSSLPinningEnabled(
        _ completion: @escaping (Bool, Error?) -> Void
    ) {
        NSLog("SSLPinningManager - Checking SSL pinning status")
        
        // Check status using UdentifySettingsProvider
        let isEnabled = UdentifySettingsProvider.isSSLPinningEnabled
        
        NSLog("SSLPinningManager - SSL pinning enabled: \(isEnabled)")
        completion(isEnabled, nil)
    }
}

